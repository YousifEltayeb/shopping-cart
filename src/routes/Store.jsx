import styled from "styled-components";
import useFilms from "../hooks/useFilms.jsx";
const Main = styled.main`
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(12rem, 1fr));

  gap: 3rem;
  padding: 1rem;
`;
const Card = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  gap: 1rem;
  background-color: #262626;
  box-shadow:
    rgba(50, 50, 93, 0.25) 0px 50px 100px -20px,
    rgba(0, 0, 0, 0.3) 0px 30px 60px -30px;
`;
const Quantity = styled.div`
  display: flex;
`;
const QuantityInput = styled.input`
  &::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
  &::-webkit-outer-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
  width: 100%;
`;
const AddToCartButton = styled.button`
  background-color: #ff0101;
  color: black;
  margin-top: auto;
`;
const Store = () => {
  const { products, setProducts, error, loading } = useFilms();
  const changeQuantity = (clickedButton) => {
    const inputFieldElm = clickedButton.parentNode.querySelector("input");
    if (clickedButton.textContent === "+") {
      inputFieldElm.value = ++inputFieldElm.value;
    } else if (clickedButton.textContent === "-" && inputFieldElm.value > 0) {
      inputFieldElm.value = --inputFieldElm.value;
    }
  };
  const addToCart = (targetElm, filmID) => {
    const inputFieldElmValue =
      targetElm.parentNode.querySelector("input").value;
    const filmIndex = products.findIndex((product) => product.id === filmID);
    if (inputFieldElmValue <= 0) {
      alert("Quantity cannot be 0");
      return;
    }
    const newProducts = [...products];
    newProducts[filmIndex].inCart = true;
    newProducts[filmIndex].quantity = Number(inputFieldElmValue);
    setProducts(newProducts);
  };
  if (loading) return <Main aria-label="Store">Loading...</Main>;
  if (error)
    return (
      <Main aria-label="Store">
        A network error was encountered <br />
        {error}
      </Main>
    );
  return (
    <Main aria-label="Store">
      {products.map((product) => {
        return (
          <Card key={product.id}>
            <img src={product.posterPath} alt="film poster" />
            <h4>{product.title}</h4>
            <Quantity>
              <button
                id="decrement"
                onClick={(e) => {
                  changeQuantity(e.target);
                }}
              >
                -
              </button>
              <QuantityInput type="number" />
              <button
                id="increment"
                onClick={(e) => {
                  changeQuantity(e.target);
                }}
              >
                +
              </button>
            </Quantity>
            <AddToCartButton
              onClick={(e) => {
                addToCart(e.target, product.id);
              }}
            >
              Add to cart
            </AddToCartButton>
          </Card>
        );
      })}
    </Main>
  );
};

export default Store;
