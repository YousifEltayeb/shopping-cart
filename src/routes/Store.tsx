import styled from "styled-components";
import useFilms from "../hooks/useFilms";
import { useProducts } from "../App";
const Main = styled.main`
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(12rem, 1fr));
  gap: 3rem;
  padding: 1rem;
`;
const Card = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  gap: 1rem;
  background-color: #262626;
  box-shadow:
    rgba(50, 50, 93, 0.25) 0px 50px 100px -20px,
    rgba(0, 0, 0, 0.3) 0px 30px 60px -30px;
`;
const Quantity = styled.div`
  display: flex;
`;
const QuantityInput = styled.input`
  &::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
  &::-webkit-outer-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
  width: 100%;
`;
const AddToCartButton = styled.button`
  background-color: #ff0101;
  color: black;
  margin-top: auto;
`;
const Store = () => {
  const { products, setProducts } = useProducts();
  const { fetchedProducts, error, loading } = useFilms();
  if (fetchedProducts) setProducts(fetchedProducts);
  const handleQuantityInc = (newQuantity: number, productId: number) => {
    handleQuantityChange(newQuantity, productId);
  };
  const handleQuantityDec = (newQuantity: number, productId: number): void => {
    handleQuantityChange(newQuantity, productId);
  };
  const handleQuantityChange = (
    newQuantity: number,
    productId: number,
  ): void => {
    const updatedProducts = [...products];
    const filmIndex = updatedProducts.findIndex(
      (product) => product.id === productId,
    );

    updatedProducts[filmIndex].quantity = newQuantity < 0 ? 0 : newQuantity;

    setProducts(() => updatedProducts);
  };
  const updateProducts = (productId: number) => {
    const updatedProducts = [...products];
    const filmIndex = updatedProducts.findIndex(
      (product) => product.id === productId,
    );
    if (updatedProducts[filmIndex].quantity > 0)
      updatedProducts[filmIndex].inCart = true;
    setProducts(() => updatedProducts);
  };
  if (loading)
    return (
      <Main data-testid="loading" aria-label="Store">
        Loading...
      </Main>
    );
  if (error)
    return (
      <Main data-testid="api-error" aria-label="Store">
        A network error was encountered <br />
        {error}
      </Main>
    );
  return (
    <Main aria-label="Store">
      {products.map((product) => {
        return (
          <Card key={product.id}>
            <img src={product.posterPath} alt="film poster" />
            <h4 data-testid="product-title">{product.title}</h4>
            <Quantity>
              <button
                id="decrement"
                data-testid="decrement-button"
                onClick={() =>
                  handleQuantityDec(product.quantity - 1, product.id)
                }
              >
                -
              </button>
              <QuantityInput
                data-testid="quantity-input"
                type="number"
                min="1"
                value={product.quantity}
                onChange={() =>
                  handleQuantityChange(product.quantity, product.id)
                }
                required
              />
              <button
                id="increment"
                data-testid="increment-button"
                onClick={() =>
                  handleQuantityInc(product.quantity + 1, product.id)
                }
              >
                +
              </button>
            </Quantity>
            <AddToCartButton
              data-testid="add-to-cart-button"
              onClick={() => updateProducts(product.id)}
            >
              Add to cart
            </AddToCartButton>
          </Card>
        );
      })}
    </Main>
  );
};

export default Store;
