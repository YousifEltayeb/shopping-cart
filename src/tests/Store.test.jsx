import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import Store from "../routes/Store";
import { Routes, Route, MemoryRouter } from "react-router";
import userEvent from "@testing-library/user-event";
import { useOutletContext } from "react-router";
// const setProducts = vi.fn();
// vi.mock("react-router", async () => ({
//   ...vi.importActual("react-router"),
//   useOutletContext: () => vi.fn(),
// }));
vi.mock("react-router");
const products = [
  {
    id: 1,
    title: "some title",
    posterPath: "some/path",
    inCart: false,
    price: 20,
    quantity: 1,
  },
];
const setProducts = vi.fn();
describe("Store", () => {
  it("Store is loading", () => {
    const emptyProducts = [];
    useOutletContext.mockReturnValue([emptyProducts]);
    const { container } = render(<Store />);

    expect(container).toMatchSnapshot();
  });

  it("Clicking add to cart and increment updates proprties correctly", async () => {
    useOutletContext.mockReturnValue([products, setProducts]);
    const user = userEvent.setup();
    render(<Store />);

    await user.click(screen.getByRole("button", { name: "+" }));
    await user.click(screen.getByRole("button", { name: "Add to cart" }));

    expect(products[0]).toEqual({ ...products[0], quantity: 1, inCart: true });
  });
  it("It doesn't update inCart or quantity");
});
