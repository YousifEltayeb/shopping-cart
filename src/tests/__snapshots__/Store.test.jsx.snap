// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Store > Store is loading 1`] = `
<div>
  <main
    aria-label="Store"
    class="sc-FEMpB cRahFd"
  >
    Loading...
  </main>
</div>
`;
